server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -2464,203 +2465,255 @@ public boolean assertAllListenersResolved() {
                     }
                 }
             }
-
-            final SnapshotsInProgress.Entry updatedEntry;
-            if (shards != null) {
-                assert clones == null : "Should not have updated clones when updating shard snapshots but saw " + clones +
-                        " as well as " + shards;
-                updatedEntry = entry.withShardStates(shards.build());
-            } else if (clones != null) {
-                updatedEntry = entry.withClones(clones.build());
+            SnapshotsInProgress.Entry updatedEntry;
+            if (shardsBuilder != null) {
+                assert clonesBuilder == null : "Should not have updated clones when updating shard snapshots but saw " + clonesBuilder +
+                        " as well as " + shardsBuilder;
+                updatedEntry = entry.withShardStates(shardsBuilder.build());
+            } else if (clonesBuilder != null) {
+                updatedEntry = entry.withClones(clonesBuilder.build());
             } else {
                 updatedEntry = entry;
             }
             entries.add(updatedEntry);
         }
-        if (changedCount > 0) {
-            logger.trace("changed cluster state triggered by [{}] snapshot state updates and resulted in starting " +
-                    "[{}] shard snapshots", changedCount, startedCount);
-            return ClusterStateTaskExecutor.ClusterTasksResult.<ShardSnapshotUpdate>builder().successes(tasks)
-                    .build(ClusterState.builder(currentState).putCustom(SnapshotsInProgress.TYPE,
-                            SnapshotsInProgress.of(entries)).build());
+
+        private ImmutableOpenMap.Builder<RepositoryShardId, ShardSnapshotStatus> clonesBuilder() {
+            assert shardsBuilder == null;
+            if (clonesBuilder == null) {
+                clonesBuilder = ImmutableOpenMap.builder(current.clones());
+            }
+            return clonesBuilder;
         }
-        return ClusterStateTaskExecutor.ClusterTasksResult.<ShardSnapshotUpdate>builder().successes(tasks).build(currentState);
-    };
+
+        private ImmutableOpenMap.Builder<ShardId, ShardSnapshotStatus> shardsBuilder() {
+            assert clonesBuilder == null;
+            if (shardsBuilder == null) {
+                shardsBuilder = ImmutableOpenMap.builder(current.shards());
+            }
+            return shardsBuilder;
+        }
+
+        /**
+         * Start shard level operation for given {@code shardId}.
+         *
+         * @param newStates   builder for updated shard states mapping
+         * @param nodeId      node id to execute started operation on
+         * @param generation  shard generation to base started operation on
+         * @param shardId     shard identifier of shard to start operation for
+         */
+        private <T> void startShard(ImmutableOpenMap.Builder<T, ShardSnapshotStatus> newStates, String nodeId, String generation,
```

Comment:
nit: maybe rename it to `startShardOperation` for clarity?

Reply:
++